### ISSUE TITLE
ServletCacheEngine ignore cache for App using default context root. 

### ISSUE THREAD

**Issue by dlxu** [OpenLiberty/open-liberty#9085]
In ServletCacheEngine.java

```java
public javax.servlet.Servlet getProxiedServlet(javax.servlet.Servlet s) {

		if (tc.isEntryEnabled())
			Tr.entry(tc, "getProxiedServlet", s);

		try {

			String name = s.getClass().getName().toLowerCase();
			if (tc.isDebugEnabled()) {
				Tr.debug(tc, "looking for ", s.getServletConfig(), name );
			}
			
			if (null != s.getServletConfig()){
				String contextRoot = s.getServletConfig().getServletContext().getContextPath(); 
				if (null != getServletCache() || !excludedServlets.contains(name)) {
					if (contextRootsWithCachespecXMLs.contains(contextRoot)) {
						s = new ServletWrapper(s);
					}
				}
			}
```

If applcation's context root is set to "/", the `servletContext.getContextPath()` will return an empty string, which does not match the `contextRoot` stores in `contextRootsWithCachespecXMLs`. And, hence the servlet caching is ignored for the application.

```
getContextPath
java.lang.String getContextPath()
Returns the context path of the web application. 
The context path is the portion of the request URI that is used to select the context of the 
request. The context path always comes first in a request URI. The path starts with a / character 
but does not end with a / character. For servlets in the default (root) context, this method returns "".
```

Please kindly fix the issue. Thanks.

**Comment by kevin-ortega** [comment]
@dlxu This is an interesting problem.  There is a cost associated with using servlet caching.  The goal of using the context root to associate a webapp with a cachespec.xml is to only insert servlet caching in the flow if the webapp's output has a reasonable chance of being cached.  If not, then we don't want the webapp to suffer the cost of servlet caching and not get any of the benefits.  If we inserted servlet caching for all webapps with a null context root then we are defeating this purpose.   Therefirem, only webapps with a context root that is associated with cachespec.xml will flow through servlet caching.

**Comment by dlxu** [comment]
Thank you for your response, Kevein @kevin-ortega 

Maybe I can describe scenario in more details here:

If in `Application.xml` or in `ibm-web-ext.xml` , we have 
``` xml
<module>
    <web>
      <web-uri>testweb.war</web-uri>
      <context-root>/</context-root>
    </web>
  </module>
```
or 
```xml
<web-ext
	xmlns="http://websphere.ibm.com/xml/ns/javaee"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://websphere.ibm.com/xml/ns/javaee http://websphere.ibm.com/xml/ns/javaee/ibm-web-ext_1_0.xsd"
	version="1.0">
             ...
	<context-root uri="/" />
               ...
</web-ext>
```

In this case, `ServletContext.getContextPath()` will return an empty string `""` not `null`, and in `contextRootsWithCachespecXMLs` we have `"/"` instead, since it is possibly read from `Application.xml` or in `ibm-web-ext.xml`.  

So, that means the entry `"/"` in `contextRootsWithCachespecXMLs`  is useless and not respected by CacheEngine.  However, in my opinion, it indeed should map to the empty string returned by `ServletContext.getContextPath()`. 

Based on what I see here, the `ServletContext.getContextPath()` is not guaranteed to be the same as the entries in `contextRootsWithCachespecXMLs`. As a result, using `String contextRoot = s.getServletConfig().getServletContext().getContextPath();`  to compare with `contextRoot` in `contextRootsWithCachespecXMLs` is probably not accurate all the time. Not sure if the Liberty team can do some preprocess of the `String contextRoot` before comparing?

There is no need to insert an additional empty string (you mentioned as `null` context root) into `contextRootsWithCachespecXMLs`, I think do some preprocess of the `contextRoot` returned by 
`s.getServletConfig().getServletContext().getContextPath();` is probably a better way.

**Comment by dlxu** [comment]
In sum, I think it is probably safe to do:
1. do not need modify the way how `contextRootsWithCachespecXMLs` is built.
2. pre-process the result returned by `s.getServletConfig().getServletContext().getContextPath()` before `contextRootsWithCachespecXMLs.constains()` test. e.g.
```java
  String contextRoot = s.getServletConfig().getServletContext().getContextPath().isEmpty() ? 
                    "/" : s.getServletConfig().getServletContext().getContextPath().isEmpty();
  ....
  if (contextRootsWithCachespecXMLs.contains(contextRoot)) {
   ....
```

@kevin-ortega What is your opinion regarding this? Thank you.

**Comment by kevin-ortega** [comment]
@dlxu I do fully understand the issue you are having.  "/" is in `contextRootsWithCachespecXMLs` but `s.getServletConfig().getServletContext().getContextPath();`  returns an empty string and therefore there will never been a match. 

I do think your code suggestion would work.  However, the problem with it is **all** web modules with a default context root would then flow through servlet caching.  We only want apps that have a reasonable chance of being cached to flow through servlet caching.  Servlet caching has a cost.  We don't want apps to suffer the cost and not get any benefit.  

This would also change the default behavior in liberty which is something we avoid as liberty has a zero migration strategy.  I suppose we could introduce this with a custom property that if enabled would provide the solution you are looking for.   Let me ponder this for a bit.

**Comment by dlxu** [comment]
@kevin-ortega 
Thank you for your explanation on this.  :)
Custom property is definitely a much better way to handle this without breaking existing behavior of Liberty.
I am looking forward to hearing your update on this.

**Comment by dlxu** [comment]
@kevin-ortega , please let me know which Liberty release/build will have this update, thanks.

**Comment by kevin-ortega** [comment]
@dlxu This will get into 19.0.0.12.  We are still discussing whether a custom property is warranted.

**Comment by dlxu** [comment]
cool, thanks.

**Comment by dlxu** [comment]
@kevin-ortega  what is the difference between open liberty and IBM Liberty?  Will the update be available in IBM Liberty also ? We are using `WebSphere Application Server Version 19.0.0.7 Liberty`

**Comment by kevin-ortega** [comment]
@dlxu Good question.  open liberty is the open-source, limited feature version of WebSphere Liberty.  If you want support with open liberty you have to purchase support.   WebSphere Liberty is built, in part, from this repo and this update will be in WebSphere Liberty in the 19.0.0.12 release.

**Comment by kevin-ortega** [comment]
@dlxu Good news.  This fix made it into 19.0.0.11. 

The estimated deliver is Nov 15:
https://www.ibm.com/support/pages/latest-fix-packs-websphere-application-server

**Comment by dlxu** [comment]
Great! Thank you! @kevin-ortega
